File: frontend/src/utils/api/api.client.ts
import axios from 'axios';
import { useAuthStore } from '../../store/useAuthStore';
import { useAlertStore } from '@/store/useAlertStore';

const apiClient = axios.create({
    baseURL: '/api',
    headers: {
        'Content-Type': 'application/json',
    },
});

apiClient.interceptors.request.use(
    (config) => {
        const { token, language } = useAuthStore.getState();
        if (token) config.headers.Authorization = `Bearer ${token}`;
        if (language) config.headers['Accept-Language'] = language;
        return config;
    },
    (error) => Promise.reject(error)
);

apiClient.interceptors.response.use(
    (response) => {
        const { method } = response.config;
        const { addAlert } = useAlertStore.getState();

        const methodsToNotify = ['post', 'put', 'patch', 'delete'];

        if (method && methodsToNotify.includes(method.toLowerCase())) {
            const message = response.data?.message;
            if (message && message !== "events.success.default") {
                addAlert(message, 'success');
            }
        }
        return response;
    },
    (error) => {
        const { addAlert } = useAlertStore.getState();
        const { logout } = useAuthStore.getState();
        const config = error.config;

        if (error.response?.status === 401) {
            addAlert("Sesión expirada", 'error');
            logout();
        } else {
            // SILENCIAR alertas en GET para evitar errores acumulados en búsqueda
            if (config?.method?.toLowerCase() !== 'get') {
                const errorMessage = error.response?.data?.message || "common.error.generic";
                addAlert(errorMessage, 'error');
            }
        }
        return Promise.reject(error);
    }
);

export default apiClient;

File: mobile/lib/services/api_client.dart
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class ApiClient {
  static const String _baseUrl = 'https://talleresmecanicos.quanticarch.com/api/v1'; 
  final http.Client _client = http.Client();

  String get baseUrl => _baseUrl;

  Future<http.Response> post(String endpoint, Map<String, dynamic> body) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('token');

    final response = await _client.post(
      Uri.parse('$_baseUrl$endpoint'),
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
      body: jsonEncode(body),
    );
    return response;
  }

  Future<http.Response> get(String endpoint) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('token');

    final response = await _client.get(
      Uri.parse('$_baseUrl$endpoint'),
      headers: {
        if (token != null) 'Authorization': 'Bearer $token',
      },
    );
    return response;
  }

  Future<http.Response> put(String endpoint, Map<String, dynamic> body) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('token');

    final response = await _client.put(
      Uri.parse('$_baseUrl$endpoint'),
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
      body: jsonEncode(body),
    );
    return response;
  }

  Future<http.Response> patch(String endpoint, Map<String, dynamic> body) async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('token');

    final response = await _client.patch(
      Uri.parse('$_baseUrl$endpoint'),
      headers: {
        'Content-Type': 'application/json',
        if (token != null) 'Authorization': 'Bearer $token',
      },
      body: jsonEncode(body),
    );
    return response;
  }
}
